ISR function doesn't exist

Is there a library that is an implied dependency?
I'm using an esp32

https://github.com/sweavo/rotary/blob/master/rotary.cpp#L31

```
c:\Users\alice\Documents\Arduino\libraries\rotary\rotary.cpp:31:4: error: expected constructor, destructor, or type conversion before '(' token
 ISR(TIMER0_COMPA_vect) 
```

--- a/README.md
+++ b/README.md
@@ -2,7 +2,8 @@
 
 This project drives the spark fun RGB rotary encoder with pushbutton and RGB LED.
 
-It was written for Uno, and contains some pretty low-level stuff that doesn't work on my ESP8266.
+It's currently tested on Arduino Uno and ESP32
+
 
 ## Install:
 
@@ -23,7 +24,7 @@ The pin names given in the library and this documentation are defined in `Pin De
   
 ### Encoder:
 
-`ROTARY_PIN_A` and `ROTARY_PIN_B` should both be interrupt pins (e.g. 2 and 3 on an Uno).
+`ROTARY_PIN_A` and `ROTARY_PIN_B` should both be interrupt pins (e.g. 2 and 3 on an Uno) and must be adjacent (e.g. use 12 and 13 on the ESP32).
 
 You don't need pullup resistors for the example because it turns on the internal pullups.
 
@@ -37,10 +38,13 @@ Connect
 * Connect R,G,B (pins 1,2,4 of the Rotary package) through 330Ohm current-limiting resistors to `PIN_NOT_RED`, `PIN_NOT_GREEN`, `PIN_NOT_BLUE` on Arduino.
 * Connect Rotary package pin 5 (+5v) to +5v. 
 
+The LEDs have not been tested on ESP32 but should be target-independent.
+
 ### Push button:
 Pin 3 of the rotary package, the push button, connects to `PIN_ROTARY_PUSH` with a 10k pulldown resistor
 
 ## Apologies
 
 * I haven't read the API guidelines yet. I expect I have committed crimes here.
-* The library uses quite a lot of resources (both interrupt pins AND a timer ISR). I welcome contributions that simplify it.
+* The library uses quite a lot of resources two interrupt pins AND a timer ISR). But it's light on compute cycles at runtime.
+

--- a/examples/rotary-demo/rotary-demo.ino
+++ b/examples/rotary-demo/rotary-demo.ino
@@ -17,18 +17,22 @@ void cbk_rotarychange( uint16_t pos )
 void setup() {
 
   // Start comms
-  Serial.begin(9600);
+  Serial.begin(115200);
   
   // ROTARY
+#ifndef ESP32
   initRGB();
+#endif
   initRotary( cbk_rotarychange );  
 }
 
 void loop() {
   delay( DELAY_CONST );
+#ifndef ESP32
   if ( digitalRead( PIN_ROTARY_PUSH ) )
     {
     Serial.println("Push");
   }
+#endif
 }
 

--- a/library.properties
+++ b/library.properties
@@ -1,5 +1,5 @@
 name=Rotary
-version=0.1.0
+version=0.2.0
 author=sweavo@gmail.com
 maintainer=Steve <sweavo@gmail.com>
 sentence=Library to drive the Spark Fun rotary encoder with pushbutton and RGB LED.

--- a/rotary.cpp
+++ b/rotary.cpp
@@ -1,6 +1,41 @@
 #include "arduino.h"
 #include "rotary.h" 
 
+#ifndef ESP32
+// ESP32 can trigger interrupts from any pin Uno only has a couple pins that can trigger interrupts
+#if ( PIN_ROTARY_A != 2 )
+#warning PIN_ROTARY_A is not 2. I hope you used a pin with interrupt support.
+#endif
+#if ( PIN_ROTARY_B != 3 )
+#warning PIN_ROTARY_B is not 3. I hope you used a pin with interrupt support.
+#endif
+#endif
+
+#if ( PIN_ROTARY_A == PIN_ROTARY_B - 1 )
+#define ROTARY_SHIFT PIN_ROTARY_A
+#elif ( PIN_ROTARY_B == PIN_ROTARY_A - 1 )
+#define ROTARY_SHIFT PIN_ROTARY_B
+#else
+#error PIN_ROTARY_A and _B must be adjacent.
+#endif
+
+#ifndef ESP32
+// On ESP32 this is an attribute needed for ISRs; it's null in Arduino
+#define IRAM_ATTR
+#endif
+
+// On ESP32, attachInterrupt just takes the pin number, but in Uno you need to call digitalPinToInterrupt
+#ifdef ESP32
+#define PIN_TO_INTERRUPT(x) x
+#else
+#define PIN_TO_INTERRUPT(x) digitalPinToInterrupt(x)
+#endif
+
+// On ESP32, you need to call REG_READ to get the input port register; not so on Uno
+#ifdef ESP32
+#define PIND REG_READ(GPIO_IN_REG)
+#endif
+
 // Given a color with bits 0, 1 and 2 representing R, G, B,
 // write to the RGB pins and change the color of the rotary
 void setRGB( int color)
@@ -24,11 +59,18 @@ void initRGB()
 // only one alarm, and a call to setAlarm before the alarm has expired results
 // in the old alarm being canceled and the new alarm set.
 ////////////////////////////////////////////////////////////////////////////////
-static unsigned long alarm_ticks = 0;
+#ifdef ESP32
+hw_timer_t * timer = NULL;
+#endif
+volatile static unsigned long alarm_ticks = 0;
 static unsigned long alarm_timeout = 0;
 static void (*alarm_callback)(void);
 
+#ifdef ESP32
+void IRAM_ATTR isr_1ms(void)
+#else
 ISR(TIMER0_COMPA_vect) 
+#endif
 {
   alarm_ticks++;
   if ( ( alarm_ticks == alarm_timeout ) && alarm_timeout )
@@ -49,12 +91,19 @@ void initAlarmService()
 {
   // Timer0 cycles at approx. 1 millisecond. We will put a CTC
   // interrupt midway through its range.
+#ifdef ESP32
+  timer=timerBegin(0, 80, true);
+  timerAttachInterrupt(timer, &isr_1ms, true);
+  timerAlarmWrite(timer, 1000, true);
+  timerAlarmEnable(timer);
+#else
   OCR0A = 0xAF; // Compare Register
   TIMSK0 |= _BV(OCIE0A); // Mask
   TCCR1B |= _BV(WGM12);  // I think this sets CTC mode
+#endif
 }
 
-void isr_rotaryupdated();
+void IRAM_ATTR isr_rotaryupdated();
 
 // Clicks of rotary switch
 static rotary_position_t rotary_position=0;
@@ -70,16 +119,15 @@ static void (*rotary_change_callback)(rotary_position_t);
 void initRotary( void (*callback)(rotary_position_t) )
 {
   rotary_change_callback = callback;
-  pinMode(PIN_ROTARY_A,INPUT_PULLUP);
-  pinMode(PIN_ROTARY_B,INPUT_PULLUP);
-  attachInterrupt( digitalPinToInterrupt( PIN_ROTARY_A ),
+  pinMode(PIN_ROTARY_A, INPUT_PULLUP);
+  pinMode(PIN_ROTARY_B, INPUT_PULLUP);
+  attachInterrupt( PIN_TO_INTERRUPT( PIN_ROTARY_A ),
                    isr_rotaryupdated,
                    CHANGE );
-  attachInterrupt( digitalPinToInterrupt( PIN_ROTARY_B ),
+  attachInterrupt( PIN_TO_INTERRUPT( PIN_ROTARY_B ),
                    isr_rotaryupdated,
                    CHANGE );
-  rotary_last_graycode = ( PIND & 12 ) >>2;
-
+  rotary_last_graycode = (PIND & (3 << ROTARY_SHIFT)) >> ROTARY_SHIFT;
   pinMode( PIN_ROTARY_PUSH, INPUT );
 
   initAlarmService();
@@ -110,11 +158,12 @@ void cbk_rotary_quiescent()
 // rotary_clicks after any bouncing has settled down.
 static const uint16_t ROTATION_TABLE[] = {  0, (ROTARY_MODULO<<2)-1, 1, 0, 1, 0,  0, (ROTARY_MODULO<<2)-1,
                                (ROTARY_MODULO<<2)-1,  0, 0, 1, 0, 1, (ROTARY_MODULO<<2)-1,  0 };
-void isr_rotaryupdated(){
-  int sample = PIND & 12;
+
+void IRAM_ATTR isr_rotaryupdated() 
+{
+  int sample = ( PIND & (3<<ROTARY_SHIFT)) >> (ROTARY_SHIFT-2);
   rotary_quarters = ( rotary_quarters + 
                       ROTATION_TABLE[ sample | rotary_last_graycode ] ) % ( ROTARY_MODULO << 2 );
   rotary_last_graycode = sample >> 2;
   setAlarm( ROTARY_TIMEOUT, cbk_rotary_quiescent );
 }
-

--- a/rotary.h
+++ b/rotary.h
@@ -1,54 +1,56 @@
 #ifndef ROTARY_H
 #define ROTARY_H
+
 /*  # Rotary encoder
  *   
- *  This project's purpose is to drive the spark fun RGB rotary encoder only.
- *  
- *  ## REFERENCE:
- *  
- *  http://cdn.sparkfun.com/datasheets/Components/Switches/EC12PLRGBSDVBF-D-25K-24-24C-6108-6HSPEC.pdf
- *  http://cdn.sparkfun.com/datasheets/Components/Switches/EC12PLRGBSDVBF-D-25K-24-24C-6108-6H.pdf
- *  
- *  ## Wiring
- *  
- *  The pin names given here are defined in `Pin Definitions` below.
- *    
- *  ### Encoder:
- *  
- *  ROTARY_PIN_A and ROTARY_PIN_B should both be interrupt pins (e.g. 2 and 3 on an Uno).
- *  
- *  We don't need pullup resistors because this sketch turns on the internal pullups.
- *  
- *  Connect
- *  * pin C directly to GND
- *  * pin A to ROTARY_PIN_A.
- *  * pin B to ROTARY_PIN_B.
- *  
- *  ### LEDs:
- *  
- *  * Connect R,G,B (pins 1,2,4 of the Rotary package) through 330Ohm current-limiting resistors to PIN_NOT_RED, PIN_NOT_GREEN, PIN_NOT_BLUE on arduino.
- *  * Connect pin 5 (+5v) to +5v. 
- *  
- *  ### Push button:
- *  Pin 3 of the rotary package, the push button, connects to PIN_ROTARY_PUSH with a 10k pulldown resistor
+ *  See https://github.com/sweavo/rotary
  */
- 
+
+
 ////////////////////////////////////////////////////////////////////////////////
-// Pin definitions.
+// BEGIN user configuration
 ////////////////////////////////////////////////////////////////////////////////
+
+// Pin definitions
+// ===============
+
 // Negative logic for LEDs,  i.e. LED is on when pin is low
+
 #define PIN_NOT_RED 13  
 #define PIN_NOT_GREEN 12
 #define PIN_NOT_BLUE 10 
+
 // Switches in rotary
+// ==================
+
 #define PIN_ROTARY_A 12
 #define PIN_ROTARY_B 13
 #define PIN_ROTARY_PUSH 11
 
+// Debouncing
+// ==========
+
+// How long must there have been no changes to the gray code before we consider
+// firing a rotary change event?
+
+#define ROTARY_TIMEOUT  4
+
+// How many unique values rotary_position can take
+
+#define ROTARY_MODULO (16384)
+
+
 ////////////////////////////////////////////////////////////////////////////////
-// Driver for RGB LEDs in rotary
+// END user configuration
 ////////////////////////////////////////////////////////////////////////////////
+
+
+////////////////////////////////////////////////////////////////////////////////
+// RGS LEDs help
+////////////////////////////////////////////////////////////////////////////////
+
 // Bitmasks for getting colors out of the RGB LEDs
+
 #define RGB_BLACK   0
 #define RGB_RED     1
 #define RGB_GREEN   2
@@ -58,30 +60,23 @@
 #define RGB_CYAN    6
 #define RGB_WHITE   7
 
+// Set the pin directions and initialize to off
+
 void initRGB();
 
 // Given a color with bits 0, 1 and 2 representing R, G, B,
 // write to the RGB pins and change the color of the rotary
+
 void setRGB( int color);
+
 
 ////////////////////////////////////////////////////////////////////////////////
 // ROTARY handling
 ////////////////////////////////////////////////////////////////////////////////
-
-// BEGIN user configuration
-
-// How long must there have been no changes to the gray code before we consider
-// firing a rotary change event?
-#define ROTARY_TIMEOUT  4
-
-// How many unique values rotary_position can take
-#define ROTARY_MODULO (16384)
-
-// END user configuration
-
 
 typedef uint16_t rotary_position_t;
 
 void initRotary( void (*callback)(rotary_position_t) );
 
 #endif /* ROTARY_H */
+

--- a/rotary.h
+++ b/rotary.h
@@ -41,10 +41,9 @@
 #define PIN_NOT_GREEN 12
 #define PIN_NOT_BLUE 10 
 // Switches in rotary
-#define PIN_ROTARY_A 2
-#define PIN_ROTARY_B 3
+#define PIN_ROTARY_A 12
+#define PIN_ROTARY_B 13
 #define PIN_ROTARY_PUSH 11
-
 
 ////////////////////////////////////////////////////////////////////////////////
 // Driver for RGB LEDs in rotary
@@ -79,13 +78,6 @@ void setRGB( int color);
 #define ROTARY_MODULO (16384)
 
 // END user configuration
-
-#if ( PIN_ROTARY_A != 2 )
-#warning PIN_ROTARY_A is not 2. I hope you used a pin with interrupt support.
-#endif
-#if ( PIN_ROTARY_B != 3 )
-#warning PIN_ROTARY_B is not 3. I hope you used a pin with interrupt support.
-#endif
 
 
 typedef uint16_t rotary_position_t;

